[Security] Fix valid remember-me token exposure to the second consequent request

zerkms opened the pull request:
Close https://github.com/symfony/symfony/issues/42343
Fix https://github.com/symfony/symfony/pull/46760

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #42343, Fix #46760 <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features -->
<!--
Replace this notice by a short README for your feature/bugfix.
This will help reviewers and should be a good start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against the latest branch.
 - For new features, provide some code snippets to help understand usage.  
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
 - Never break backward compatibility (see https://symfony.com/bc).
-->

https://github.com/symfony/symfony/pull/46760 PR together with a fix produces a security vulnerability when a malicious actor may get a **new and valid** remember me token if makes a request right after the legit user.

nicolas-grekas reviewed (approved):
This change looks correct to me.

2 requests come by at the same time with tokenA:
- the one that wins the race persists and sends back tokenB
- the other one accepts tokenA but doesn't send any cookie back, aka tokenB from req 1 stays in the cookie jar.

Unless I missed something, this still fixes the situation that @heiglandreas describes in #46760 while preventing a needless disclosure of the currently valid token to the 2nd request.

/cc @Seldaek in case you want to have a look.

I'm just worried that although this causes a possible security threat, [the security disclosure process](https://symfony.com/security) hasn't been followed appropriately. Please be more careful in the future.

nicolas-grekas commented on src/Symfony/Component/Security/Http/Tests/RememberMe/PersistentRememberMeHandlerTest.php:
```diff
@@ -126,17 +126,7 @@ public function testConsumeRememberMeCookieValidByValidatorWithoutUpdate()
         $handler->consumeRememberMeCookie($rememberMeDetails);
 
         // assert that the cookie has been updated with a new base64 encoded token value
```
can you please update this comment? (or remove it)

zerkms replied:
Totally, I seen it but totally forgot to remove, apologies.

Commit:
Bug #42343 [Security] Fix valid remember-me token exposure to the second consequent request

Close https://github.com/symfony/symfony/issues/42343
Fix https://github.com/symfony/symfony/pull/46760

nicolas-grekas commented:
Thank you @zerkms.

heiglandreas commented:
Throughout the last days I debugged some odd occurrences and was able to track them down to this PR.

I will create a new issue for that so that we can track that accordingly.
